### PR TITLE
Support reading log-only Hudi MOR table

### PR DESCRIPTION
**Pull request by choucmei**
support read log-only on hudi MOR table

Ref: https://github.com/prestodb/presto/issues/17973

Test plan - Unit tests.

```
== RELEASE NOTES ==

Hudi Change
* support read log-only on hudi MOR table

Hive Changes
* add hudi has log file only IntegrationTests ,and test data in zip package
```

### CODE REVIEW

**Review by 7c00 — CHANGES_REQUESTED** [review]
Could you add some tests for this change? You can take TestIntegrationTests as an example. The minor diff is to create a tiny MOR table with only log files.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java
Comment (7c00) [inline presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java:147]: A better way to handle split generation:

- change return type of createHudiSplit to `Optional<HudiSplit>`
-  if `table` type is COW and `slice` has no base file, return an empty `Optinal<HudiSplit>`
- filter empty optionals before `builder::add` in Line125.
- `HiveEmptySplitPageSource` is not needed any longer.

This might make more sense. Because we don't need to dispatch an empty split.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java
Comment (7c00) [inline presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java:156]:
```suggestion
        Optional<HudiFile> hudiFile = slice.getBaseFile().map(f -> new HudiFile(f.getPath(), 0, f.getFileLen());
        if (hudiFile.isEmpty() && table.getTableType() == HudiTable.COW) {
             return Optional.empty(); 
        }
```

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java
Comment (7c00) [inline presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java:164]: If the split have no base file, we should not put a log file into HudiSplit#baseFile field. This could add misleading  when we will create a new MOR reader to handle the file slice that has no base file.  Just put `hudFile` I suggested above.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java
Comment (7c00) [inline presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java:102]: After we change HudiSplitManager, we have confident that hudiSplit.getBaseFile().isPresent() is always true. Add an assertion here. We dont need HiveEmptySplitPageSource

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java
Comment (7c00) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java:164]: For now, we could take use of the first log file as base file in HudiRecordCursors.createRealtimeRecordCursor.

**Review by 7c00 — CHANGES_REQUESTED** [review]

File: presto-hive/src/test/java/com/facebook/presto/hive/hudi/TestHudiIntegration.java
Comment (7c00) [inline presto-hive/src/test/java/com/facebook/presto/hive/hudi/TestHudiIntegration.java:67]: why do we need to append a blank string?

File: presto-hive/src/test/java/com/facebook/presto/hive/hudi/TestHudiIntegration.java
Comment (7c00) [inline presto-hive/src/test/java/com/facebook/presto/hive/hudi/TestHudiIntegration.java:113]: nit: -> sqlResultEmpty

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java
Comment (7c00) [inline presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java:139]: Fix to follow code style.

**Review by 7c00 — CHANGES_REQUESTED** [review]
The PR is almost perfect. Following is the remaining actions to do before merging:

1. Fix the code style issues, see [Presto-Development-Guidelines#formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting)
2. Rebase your commits into a single commit
3. Refine the PR title (as well as your commit message), eg. `Support reading log-only Hudi table`
4. Refine the PR description, add a proper release note, see [Release-Notes-Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines)

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java
Comment (choucmei) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java:162]: > Let's put `hudiFile` as it is.
> 
> Move the logic of using the first log file as base file for the MOR table into `HudiPageSourceProvider`.

HudiSplit#baseFile is a final variable,Move the logic into HudiPageSourceProvider need to create new object of HudiSplit or add set method and remove `final`. It is not graceful.

Comment (7c00) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java:162]: Creating a new object is fine. Don't worry.

**Review by 7c00 — APPROVED** [review]
LGTM generally. 

Let's complete the PR by finish the remaining tasks:

1. Resolve my new comment ;
2. Rebase your branch with the latest master branch;
3. Refine your commit message; it's good to use the PR title;
4. Refine the PR description: please follow the format as other merged PR , eg. https://github.com/prestodb/presto/pull/17149#issue-1091484780

**Comment by 7c00** [comment]
Hi, @tdcmeehan could you take another review on this PR?

**Review by tdcmeehan — APPROVED** [review]

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java
Comment (choucmei) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java:127]: `
HudiFile baseFile = split.getBaseFile().orElseThrow(
        () -> new PrestoException(HUDI_CANNOT_OPEN_SPLIT, "Split without base file not supported yet"));
Path path = new Path(baseFile.getPath());
`
it needs to be replaced twice. 

https://github.com/prestodb/presto/blob/d83b10737eab1b448c09f11061ce195947ad3666/presto-hudi/src/main/java/com/facebook/presto/hudi/HudiRecordCursors.java#L74

https://github.com/prestodb/presto/blob/d83b10737eab1b448c09f11061ce195947ad3666/presto-hudi/src/main/java/com/facebook/presto/hudi/HudiRecordCursors.java#L112

Maybe it can be avoided when building, or keep current

Comment (7c00) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java:127]: Got it. We could extract a method to reduce the repeated code.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java
Comment (vinothchandar) [inline presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java:92]: is "Split without base file is invalid for COW table type" a better message?

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java
Comment (vinothchandar) [inline presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java:118]: I find the use of the first log file as a base file, bit obtuse. Given the baseFile is an Optional anyway. Can we avoid this.

**Review by vinothchandar — CHANGES_REQUESTED** [review]
Wonder if we can find a better way to support this.

File: presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java
Comment (choucmei) [reply presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSourceProvider.java]: because hudi check log-only  by basePath
https://github.com/apache/hudi/blob/aaddaf524a07daa1ffad6d6f5e04f6b130e5b069/hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/HoodieParquetRealtimeInputFormat.java#L69

**Review by vinothchandar — APPROVED** [review]
Thanks for getting back. Looks like this goes deeper than the connector. 

Could we file a Presto issue to track fixing the helper and also raise a Hudi JIRA to get this done better in the upcoming 0.12 release? cc @codope 

Then we can land this to get things working for now.

**Comment by vinothchandar** [comment]
Some tests are failing though? It seemed unrelated. Kicked the test again now

**Comment by 7c00** [comment]
@choucmei let's squash the 3 commits into one. I will help merge this PR after that.

**Pull request merged**
